src/com/facebook/buck/go/GoListStep.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2017-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.go;
+
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.shell.ShellStep;
+import com.facebook.buck.step.ExecutionContext;
+import com.facebook.buck.util.ProcessExecutor.Option;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class GoListStep extends ShellStep {
+  enum FileType {
+    GoFiles,
+    CgoFiles,
+    SFiles,
+    HFiles,
+    TestGoFiles,
+    XTestGoFiles
+  }
+
+  private final GoToolchain goToolchain;
+  private final List<FileType> fileTypes;
+  private final GoPlatform platform;
+
+  public GoListStep(
+      BuildTarget buildTarget,
+      Path workingDirectory,
+      GoToolchain goToolchain,
+      GoPlatform platform,
+      List<FileType> fileTypes) {
+    super(Optional.of(buildTarget), workingDirectory);
+    this.goToolchain = goToolchain;
+    this.platform = platform;
+    this.fileTypes = fileTypes;
+  }
+
+  @Override
+  protected ImmutableList<String> getShellCommandInternal(ExecutionContext context) {
+    ImmutableList.Builder<String> commandBuilder =
+        ImmutableList.<String>builder()
+            .add(goToolchain.getGoRoot().resolve("bin").resolve("go").toString())
+            .add("list")
+            .add("-f")
+            .add(
+                String.join(
+                    ":",
+                    fileTypes
+                        .stream()
+                        .map(fileType -> "{{join ." + fileType.name() + " \":\"}}")
+                        .collect(Collectors.toList())));
+
+    return commandBuilder.build();
+  }
+
+  @Override
+  public String getShortName() {
+    return "go list";
+  }
+
+  @Override
+  protected void addOptions(ImmutableSet.Builder<Option> options) {
+    super.addOptions(options);
+    options.add(Option.EXPECTING_STD_OUT);
+  }
+
+  @Override
+  public ImmutableMap<String, String> getEnvironmentVariables(ExecutionContext context) {
+    return ImmutableMap.<String, String>builder()
+        // The GOROOT is set to /usr/local/go somehow on Travis CI, which doesn't exist.
```

Comment:
This comment is out of place here. Why do we need to know about Travis CI in this file?

I think it's totally reasonable to set these variables before running go command without explanation, so you can remove this comment or change it to state that the go commands rely on these variables and we set them to override external variables.